core/src/main/java/de/danoeh/antennapod/core/feed/LocalFeedUpdater.java

```diff
@@ -0,0 +1,74 @@
+package de.danoeh.antennapod.core.feed;
+
+import android.content.Context;
+import android.media.MediaMetadataRetriever;
+import android.net.Uri;
+
+import java.io.File;
+import java.io.FilenameFilter;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+import de.danoeh.antennapod.core.storage.DBTasks;
+
+public class LocalFeedUpdater {
+
+    private static long getFileDuration(File f) {
+        MediaMetadataRetriever mediaMetadataRetriever = new MediaMetadataRetriever();
+        mediaMetadataRetriever.setDataSource(f.getAbsolutePath());
+        String durationStr = mediaMetadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION);
+        return Long.parseLong(durationStr);
+    }
+
+    /** Starts the import process. */
+    public static void startImport(Uri uri, Context context) {
+        File f = new File(uri.getPath());
+        if (!f.isDirectory()) {
+            throw new RuntimeException("invalid path");
+        } else {
+            startImportDirectory(uri, context);
+        }
+    }
+
+    private static void startImportDirectory(Uri uri, Context context) {
+        File f = new File(uri.getPath());
+        String dirUrl = uri.toString();
+        Feed dirFeed = new Feed(dirUrl, null, "Local directory (" + dirUrl + ")");
+        List<FeedItem> items = new ArrayList<>();
+
+        //find relevant files and create items for them
+        File[] itemFiles = f.listFiles(
+                new FilenameFilter() {
+                    @Override
+                    public boolean accept(File file, String s) {
+                        return s.endsWith(".mp3");
+                    }
+                });
+        for (File it: itemFiles) {
+            FeedItem item = createFeedItem(it, dirFeed);
+            items.add(item);
+        }
+        dirFeed.setItems(items);
+
+        //add or merge to the db
+        Feed[] feeds = DBTasks.updateFeed(context, dirFeed);
+    }
+
+    private static FeedItem createFeedItem(File f, Feed whichFeed) {
+        //create item
+        long globalId = 0;
+        Date date = new Date();
+        FeedItem item = new FeedItem(globalId, f.getName(), "item" + Long.toString(date.getTime()), //XXX possible problem with date resolution
```

Comment:
You can use `UUID.randomUUID().toString()` as ID to make it unique.

> https://stackoverflow.com/a/3600625/4193263
> for there to be a one in a billion chance of duplication, 103 trillion version 4 UUIDs must be generated.

Reply:
Right, UUIDs are indeed better. But first we need to check whether the feeditem already exists in the database, and if so skip this file. I saw there is a search method in DBTasks, but I'm not sure how to use it [here].

Reply:
If you add the following before looping over the files, you can just use `whichFeed.getItems()` to see if the item already exists.

```
dirFeed.setItems(new ArrayList<>()); // Does not delete non-existing items
dirFeed = DBTasks.updateFeed(context, dirFeed)[0];

for (File it: itemFiles) {
    FeedItem item = createFeedItem(it, dirFeed);
    ...
```

This is kind of a workaround currently but when splitting this into two methods, it's a good idea to return the updated `Feed` instance in `Feed createFeed(…` anyway, as it contains the correct ID.